### PR TITLE
tests and fixes for gda / vrgda

### LINKED ISSUE #257: dojo-defi: implement VRGDA/GDA
Gradual dutch auctions (GDA) and variable rate gradual dutch auctions (VRDGA) are mechanisms introduced by Paradigm that help with sales of NFTs over time

Resources:
- https://www.paradigm.xyz/2022/08/vrgda
- https://www.paradigm.xyz/2022/04/gda
- Python analysis: https://github.com/FrankieIsLost/gradual-dutch-auction/blob/master/analysis/analysis.ipynb

Cairo implementations:
- https://github.com/sambarnes/cairo-dutch
- https://github.com/cartridge-gg/cairo-gda

Check: https://github.com/dojoengine/dojo/tree/main/crates/dojo-defi

### PR DESCRIPTION
**Pull request by ametel01**
impl discrete and continous gda

added gda test + fixes

Test on gda ok, need to test vrgda

closes #257

### CODE REVIEW

File: crates/dojo-defi/src/tests/discrete_gda_test.cairo [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        initial_price: FixedTrait::new_unscaled(1000, false),
+        scale_factor: div(FixedTrait::new_unscaled(11, false), FixedTrait::new_unscaled(10, false)),
+        decay_constant: div(FixedTrait::new_unscaled(1, false), FixedTrait::new_unscaled(2, false)),
+        auction_start_time: 0,
+    };
+    let price: Fixed = auction.purchase_price(1);
+    assert(price == auction.initial_price, 'wrong price')
+}
+
+// ipynb with calculations at https://colab.research.google.com/drive/14elIFRXdG3_gyiI43tP47lUC_aClDHfB?usp=sharing
```
Comment (ponderingdemocritus) [inline crates/dojo-defi/src/tests/discrete_gda_test.cairo:23]: can you make this public please!

**Comment by ponderingdemocritus** [comment]
this is looking great!

**Comment by ponderingdemocritus** [comment]
also, can you write some detailed natspec to help new people understand. This is going to be a heavily used module!

**Comment by ponderingdemocritus** [comment]
We also need to implement

VRGDA - Linear
VRGDA - logistic 
https://www.paradigm.xyz/2022/08/vrgda

But we can do this in another PR

**Comment by ametel01** [comment]
> We also need to implement
> 
> VRGDA - Linear VRGDA - logistic https://www.paradigm.xyz/2022/08/vrgda
> 
> But we can do this in another PR

I should be done with VRGDAs by today

**Comment by ametel01** [comment]
@ponderingdemocritus using raw timestamp causes overflow on exp() calculations, we need to scale it. Considering the use cases is it more appropriate to scale the timestamp on hours or days?

**Comment by ponderingdemocritus** [comment]
I would say hours is appropriate - but some devs might like days. Maybe generalise it so it's easy to choose?

**Comment by ametel01** [comment]
> also, can you write some detailed natspec to help new people understand. This is going to be a heavily used module!

wold not be rustdoc more appropiate?

**Comment by ponderingdemocritus** [comment]
same style comments as in the World.cairo

**Comment by ametel01** [comment]
Might need to add some more tests

**Comment by ametel01** [comment]
I will also  add a readme with detailed explanation

**Pull request merged**